### PR TITLE
Refactored internal and external wms layers tables into one.

### PR DESCRIPTION
**Pull request by ochriste**
Merge LayerInternalWMS LayerExternalWMS into LayerWMS (+ ServerOGC), with Alembic migration script and FormAlchemy admin interface update.

themes json example visible here:
http://geomapfish-demo.camptocamp.net/ochdev/wsgi/themes?version=2

this PR doesnt include cleaning of deprecated LayerInternalWMS and LayerExternalWMS,
I would like to do that in a separate PR

### CODE REVIEW

File: c2cgeoportal/scaffolds/update/CONST_alembic/main/versions/116b9b79fc4d_internal_and_external_layer_tables_.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+
+from alembic import op, context
+from sqlalchemy import ForeignKey, Column, Table, MetaData
+from sqlalchemy.types import Integer, Boolean, Unicode, String, Float, \
+    UserDefinedType, DateTime, Enum
+
+# revision identifiers, used by Alembic.
+revision = '116b9b79fc4d'
+down_revision = 'a4f1aac9bda'
+branch_labels = None
```
Comment (sbrunner) [inline c2cgeoportal/scaffolds/update/CONST_alembic/main/versions/116b9b79fc4d_internal_and_external_layer_tables_.py]: => master

File: c2cgeoportal/scaffolds/update/CONST_alembic/main/versions/116b9b79fc4d_internal_and_external_layer_tables_.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    # ocg for externals
+    # using temporary original_id column
+    op.execute(
+        "INSERT INTO %(schema)s.server_ogc (name, url, image_type, auth, is_single_tile, original_id) "
+        "SELECT 'source for ' || layer, url, image_type, 'none', "
+        "CASE WHEN is_single_tile IS TRUE THEN TRUE ELSE FALSE END, id "
+        "FROM %(schema)s.layer_external_wms" % {
+            'schema': schema,
+        }
+    )
```
Comment (sbrunner) [inline c2cgeoportal/scaffolds/update/CONST_alembic/main/versions/116b9b79fc4d_internal_and_external_layer_tables_.py]: We shouldn't have many different entries for same: url, image_type and is_single_tile ...

Comment (ochriste) [inline c2cgeoportal/scaffolds/update/CONST_alembic/main/versions/116b9b79fc4d_internal_and_external_layer_tables_.py]: I see what you mean, but I dont know how to do that

Comment (ochriste) [inline c2cgeoportal/scaffolds/update/CONST_alembic/main/versions/116b9b79fc4d_internal_and_external_layer_tables_.py]: add a distinct clause on the 3 column in the select?

File: c2cgeoportal/scaffolds/update/CONST_alembic/main/versions/116b9b79fc4d_internal_and_external_layer_tables_.py (hunk 1 above)
Comment (ochriste) [inline c2cgeoportal/scaffolds/update/CONST_alembic/main/versions/116b9b79fc4d_internal_and_external_layer_tables_.py]: ERROR [alembic.util] Branch name 'master' in revision a4f1aac9bda already used by revision 116b9b79fc4d
  FAILED: Branch name 'master' in revision a4f1aac9bda already used by revision 116b9b79fc4d

**Comment by ochriste** [comment]
reste a faire:
mettre a jour 
c2cgeoportal/tests/functional/test_entry.py
c2cgeoportal/tests/functional/test_themes_time.py

**Comment by sbrunner** [comment]
Documentation also: https://camptocamp.github.io/c2cgeoportal/master/developer/webservices.html#result-in-v2

File: c2cgeoportal/scaffolds/update/CONST_alembic/main/versions/116b9b79fc4d_internal_and_external_layer_tables_.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+        "WHERE lew.image_type IS NULL AND so.image_type='image/png'" % {
+            'schema': schema,
+        }
+    )
+
+    # ocg for externals
+    op.execute(
+        'INSERT INTO %(schema)s.server_ogc (name, url, image_type, auth, is_single_tile) '
+        "SELECT 'source for ' || url, url, image_type, 'none', CASE WHEN is_single_tile "
+        "IS TRUE THEN TRUE ELSE FALSE END as is_single_tile "
```
Comment (sbrunner) [inline c2cgeoportal/scaffolds/update/CONST_alembic/main/versions/116b9b79fc4d_internal_and_external_layer_tables_.py]: `IS TRUE THEN TRUE ELSE FALSE` what is it for? for null value?

File: c2cgeoportal/scripts/themev1tov2.py [hunk 4]
```diff
@@ -63,27 +63,84 @@ def main():  # pragma: no cover
...
+        print "emptying table %s" % str(table.__table__)
+        # must be done exactly this way othewise the cascade config in the
+        # models are not used
+        for t in session.query(table).all():
+            session.delete(t)
+
+    # list and create all distinct server_ogc
+    server_ogc(session)
+
+    print "converting layerv1"
```
Comment (sbrunner) [inline c2cgeoportal/scripts/themev1tov2.py]: shouldn't be removed?

File: c2cgeoportal/scripts/themev1tov2.py [hunk 5]
```diff
@@ -63,27 +63,84 @@ def main():  # pragma: no cover
...
 
     session = DBSession()
-    session.execute(LayerInternalWMS.__table__.delete())
-    session.execute(LayerExternalWMS.__table__.delete())
-    session.execute(LayerWMTS.__table__.delete())
 
-    for layer in DBSession.query(LayerV1).all():
+    table_list = [LayerWMTS, LayerWMS, ServerOGC]
+    for table in table_list:
+        print "emptying table %s" % str(table.__table__)
```
Comment (sbrunner) [inline c2cgeoportal/scripts/themev1tov2.py]: User message should start with a capital and ends with a dot ...

File: c2cgeoportal/tests/functional/test_themes.py [hunk 6]
```diff
@@ -282,11 +290,11 @@ def test_min_levels(self):
         themes = entry.themes()
         self.assertEquals(self._get_filtered_errors(themes), set([
             u"The Layer '__test_theme/__test_layer_group_1/__test_layer_internal_wms' is under indented (1/2).",
-            u"Layer '__test_layer_external_wms' cannot be in the group '__test_layer_group_1' (internal/external mix).",
-            u"Layer '__test_layer_wmts' cannot be in the group '__test_layer_group_1' (internal/external mix).",
-            u"Layer '__test_layer_wmts' cannot be in the group '__test_layer_group_2' (internal/external mix).",
+            u"The Layer '__test_theme/__test_layer_group_1/__test_layer_wmts' is under indented (1/2).",
+            u"The Layer '__test_theme/__test_layer_group_2/__test_layer_external_wms' is under indented (1/2).",
```
Comment (sbrunner) [inline c2cgeoportal/tests/functional/test_themes.py]: Those messages looks strange it seem that the indent incrementation don't works anymore!

Comment (sbrunner) [inline c2cgeoportal/tests/functional/test_themes.py]: I just have a look on it again and finally it's OK :-)

File: c2cgeoportal/scaffolds/update/CONST_alembic/main/versions/116b9b79fc4d_internal_and_external_layer_tables_.py (hunk 3 above)
Comment (ochriste) [inline c2cgeoportal/scaffolds/update/CONST_alembic/main/versions/116b9b79fc4d_internal_and_external_layer_tables_.py]: for NULL yes

File: c2cgeoportal/scripts/themev1tov2.py (hunk 4 above)
Comment (ochriste) [inline c2cgeoportal/scripts/themev1tov2.py]: mmm, without feedback, the user doesn't know whats being done, so I though a bit of text would be welcome, but it's not a necessity.

**Comment by ochriste** [comment]
reste a faire:

documentation / schéma
 https://camptocamp.github.io/c2cgeoportal/master/developer/webservices.html#result-in-v2

File: doc/developer/webservices.rst [hunk 7]
```diff
@@ -74,24 +80,39 @@ Layer:
 
     {
         "name": <name>,
-        "type": "internal WMS/external WMS/WMTS",
-        "ui_metadata": {
+        "type": "<map server type>/WMTS/null",
```
Comment (sbrunner) [inline doc/developer/webservices.rst]: What's `<map_server_type>`?

File: doc/developer/webservices.rst [hunk 8]
```diff
@@ -74,24 +80,39 @@ Layer:
...
-        "layer": <wms_layers>,
-        "image_type": "image/png",
-        "style": <style>,
-        "queryable": 0/1,
+
+        "layers": <wms_layers>,
+        "name": <name in tree>,
         "minResolutionHint": <minResolutionHint>,
         "maxResolutionHint": <maxResolutionHint>,
+        "url": <wms server url>/null,
```
Comment (sbrunner) [inline doc/developer/webservices.rst]: It can be null?

File: doc/developer/webservices.rst [hunk 9]
```diff
@@ -74,24 +80,39 @@ Layer:
...
+        "url_wfs": <wfs server url>/null,
```
Comment (sbrunner) [inline doc/developer/webservices.rst]: null or undefined?

Comment (ochriste) [inline doc/developer/webservices.rst]: null

File: doc/developer/webservices.rst (hunk 8 above)
Comment (ochriste) [inline doc/developer/webservices.rst]: if the layer is an internal wms, yes

File: doc/developer/webservices.rst [hunk 10]
```diff
@@ -22,6 +22,11 @@ Parameters
 * ``min_levels``: minimum number of group levels that's required, default is ``1``.
 * ``role``: role name, not used by the server but it's required for the cache management.
 
+Requests examples:
+
+* themes?version=2
+* themes?version=2&background_layers_group=background
```
Comment (sbrunner) [inline doc/developer/webservices.rst]: it's not `themes?version=2&background=background`?

**Comment by sbrunner** [comment]
Just a last comments also that's looks good :-)

File: doc/developer/webservices.rst (hunk 10 above)
Comment (ochriste) [inline doc/developer/webservices.rst]: you are right!

**Pull request merged**